src/MaxwellSolvers/FDTD.h

```diff
@@ -0,0 +1,756 @@
+            A_np1.fillHalo();
+        }
+        void evaluate_EB(){
+            ippl::Vector<scalar, 3> inverse_2_spacing = ippl::Vector<scalar, 3>(0.5) / hr_m;
+            const scalar idt = scalar(1.0) / dt;
+            auto A_np1 = this->A_np1.getView(), A_n = this->A_n.getView(), A_nm1 = this->A_nm1.getView();
+            auto source = Maxwell<EMField, FourField>::source_mp->getView();
+            auto Eview  = Maxwell<EMField, FourField>::En_mp->getView();
+            auto Bview  = Maxwell<EMField, FourField>::Bn_mp->getView();
+
+            Kokkos::parallel_for(this->A_n.getFieldRangePolicy(), KOKKOS_LAMBDA(size_t i, size_t j, size_t k){
+                ippl::Vector<scalar, 3> dAdt = (A_n(i, j, k).template tail<3>() - A_nm1(i, j, k).template tail<3>()) * idt;
+                ippl::Vector<scalar, 4> dAdx = (A_n(i + 1, j, k) - A_n(i - 1, j, k)) * inverse_2_spacing[0];
+                ippl::Vector<scalar, 4> dAdy = (A_n(i, j + 1, k) - A_n(i, j - 1, k)) * inverse_2_spacing[1];
+                ippl::Vector<scalar, 4> dAdz = (A_n(i, j, k + 1) - A_n(i, j, k - 1)) * inverse_2_spacing[2];
+
+                ippl::Vector<scalar, 3> grad_phi{
+                    dAdx[0], dAdy[0], dAdz[0]
+                };
+                ippl::Vector<scalar, 3> curlA{
+                    dAdy[3] - dAdz[2],
+                    dAdz[1] - dAdx[3],
+                    dAdx[2] - dAdy[1],
+                };
+                Eview(i,j,k) = -dAdt - grad_phi;
+                Bview(i,j,k) = curlA;
+            });
+            Kokkos::fence();
+        }
+
+
+        bool periodic_bc;
+        
+        typename FourField::Mesh_t* mesh_mp;
+        FieldLayout<Dim>* layout_mp;
+        NDIndex<Dim> domain_m;
+        Vector_t hr_m;
+        
+        
+        Vector<int, Dim> nr_m;
+        
+        void initialize() {
+        // get layout and mesh
+            layout_mp = &(this->source_mp->getLayout());
+            mesh_mp   = &(this->source_mp->get_mesh());
+            
+            // get mesh spacing, domain, and mesh size
+            hr_m     = mesh_mp->getMeshSpacing();
+            dt = hr_m[0] / 2;
+            for (unsigned int i = 0; i < Dim; ++i){
+                dt = std::min(dt, hr_m[i] / 2);
+            }
+            domain_m = layout_mp->getDomain();
+            for (unsigned int i = 0; i < Dim; ++i)
+                nr_m[i] = domain_m[i].length();
+
+            // initialize fields
+            A_nm1.initialize(*mesh_mp, *layout_mp);
+            A_n.initialize(*mesh_mp, *layout_mp);
+            A_np1.initialize(*mesh_mp, *layout_mp);
+
+            A_nm1 = 0.0;
+            A_n   = 0.0;
+            A_np1 = 0.0;
+        };
+    };
+    /**
+     * @brief Nonstandard Finite-Difference Time-Domain
+     * 
+     * @tparam EMField 
+     * @tparam FourField 
+     */
+    template <typename EMField, typename FourField, fdtd_bc boundary_conditions = periodic>
+    class NonStandardFDTDSolver : Maxwell<EMField, FourField> {
```

Comment:
Would it make sense to make this inherit from the FDTDSolver since a lot of the stuff seems to be the same and override the step() method to change it?